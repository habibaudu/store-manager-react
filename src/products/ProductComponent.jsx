import React from 'react';
import SlideShowComponent from './SlideShowComponent';
import img2 from  '../img/walltims.jpeg';

class ProductComponent extends React.Component {

   componentDidMount(){
     this.props.fetchAllProducts();
    }
 
    handleSearch = event =>{
      const { searchProduct } = this.props;
      const product = event.target.elements.search.value.trim();
      event.preventDefault();
      searchProduct(product)
  }

  extract = img => {
    const getsrc = img
      const getalt = img
      const regex2 = /<img.*?alt='(.*?)'/;
      const regex = /<img.*?src='(.*?)'/;
      const altvalue= regex2.exec(getalt)[1]
      const srcvalue = regex.exec(getsrc)[1];
      return <img src={srcvalue} alt ={altvalue} />;
  }

    render(){
      //
      
   return (
   <div>
     <SlideShowComponent />
<div className ='mobile_header'>
            <img src ={img2} />
</div> 
<section id="title"> 
 Search for Product
 <form id='search' className='byid' onSubmit={this.handleSearch}>
    <input type="text" placeholder="Search product by name.." name="search" id ='filter2' />
    <button type="submit">search </button>
  </form>  
</section>
     
    {this.props.product ? (
      
      <div id ="aproduct" className="article2 clearfix">
                              
                             <section>
                             {this.extract(this.props.product.data.imgs)}
                             </section>
                             <section>
                             <span className ='note'> Details</span>
                              <hr />
                              <span className='note2'>productId : {this.props.product.data.id}</span><br />
                             <span className='note2'>product Name : {this.props.product.data.productname}</span><br />
                             <span className='note'>Price : #{this.props.product.data.price}</span><br />
                             <span className='note2'>Quantity : {this.props.product.data.quantity} in stock</span><br />
                              <span className='note2'>Minimum  : {this.props.product.data.minimum} Allowed in stock</span><br />
                              <span className='note2'>Date Added : {this.props.product.data.created_date} </span><br />
                              <hr /><br />
                              <span className='note'>Description : {this.props.product.data.description} </span><br />
                             </section>
                              
                             
    </div>
    ):' '}
<section id="title">
   <a href ='#'> products </a> 
   </section>

<div id ="allproduct" className="article clearfix">

{this.props.payload ? (
              this.props.payload.map( product => {
                const { productname, id,item,price,quantity} = product;
                const getsrc = item;
                const getalt = item;
                const regex2 = /<img.*?alt='(.*?)'/;
                const regex = /<img.*?src='(.*?)'/;
                const altvalue= regex2.exec(getalt)[1]
                const srcvalue = regex.exec(getsrc)[1]; 
                      
                return (
                  <div key={productname}>
                  <section>
                     {<img src={srcvalue} alt ={altvalue} />}
                     <span className='note2'>{id }</span><br/>
                     <span className='note2'>{productname}</span><br/>
                     <span className='note'>#{price}</span><br/>
                     <span className='note2'>{quantity} in stock</span><br/>
                     <button onClick ={() => {
                       this.props.addToCart(product)
                       
                     }} className='bt'>Add to Cart ({( this.props.cart.length && this.props.cart.filter(cartItem => cartItem.id === product.id).length && this.props.cart.filter(cartItem => cartItem.id === product.id)[0].quantityInCart) ||0 })</button>
                     {this.props.cart.length ?
                     <button onClick ={() => {
                       this.props.removeFromCart(this.props.cart.filter(cartItem => cartItem.id === product.id)[0])
                       
                     }} className='bt'>Remove</button> : null}
                    </section>
                  </div>
                );
              })
            ) : (
              <h3>loading product please wait...</h3>
            )}
    
     
</div>

   </div>);
}
};

export default ProductComponent;