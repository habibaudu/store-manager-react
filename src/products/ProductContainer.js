import { connect } from 'react-redux';
import { fetchAllProducts, searchProduct } from './duck';
import ProductComponent  from './ProductComponent';


const mapStateToProps = (state) => {
  return {
          payload :state.productReducer.payload,
          product: state.productReducer.product
        }
};

const mapDispatchToProps = dispatch => {
  return {
    fetchAllProducts: () =>
      dispatch(fetchAllProducts()),
    searchProduct: (product) =>
      dispatch(searchProduct(product)),
  };
};
const ProductContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProductComponent);

export { ProductContainer, mapDispatchToProps, mapStateToProps };
