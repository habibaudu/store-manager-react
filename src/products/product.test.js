import React from 'react';
import { shallow, mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Redirect, MemoryRouter } from 'react-router-dom';
import axios from 'axios';
import ProductComponent from './ProductComponent';
import { ProductContainer, mapDispatchToProps } from './ProductContainer';
import {
  actions,
  types,
  constants,
  productReducer,
  fetchAllProducts,
  searchProduct,
} from './duck';

jest.mock('axios');

const { setAllProducts, getAProducts, addToCart, removeFromCart } = actions;
const props = {
  fetchAllProducts: jest.fn(),
  searchProduct: jest.fn(),
  addToCart: jest.fn(),
  removeFromCart: jest.fn(),

  cart: {
    data: [
      {
        productname: 'shoe lace',
        quantity: 3,
        minimum: 4,
        description: 'a good product',
        price: 230,
      },
      {
        productname: 'shoe lace2',
        quantity: 33,
        minimum: 4,
        description: 'a good product',
        price: 230,
      },
    ],
  },
};

describe('PRODUCT TEST SUITE', () => {
  describe('Product Component', () => {
    it('should render the PRODUCT Page', () => {
      const component = shallow(<ProductComponent {...props} />);
      expect(component.exists()).toBe(true);
      expect(component).toMatchSnapshot();
    });
    it('it iterate through product', () => {
      const props = {
        fetchAllProducts,
        payload: [
          {
            productname: 'shoe lace',
            quantity: 3,
            minimum: 4,
            description: 'a good product',
            price: 230,
          },
          {
            productname: 'shoe lace2',
            quantity: 33,
            minimum: 4,
            description: 'a good product',
            price: 230,
          },
        ],

        cart: {
          data: [
            {
              productname: 'shoe lace',
              quantity: 3,
              minimum: 4,
              description: 'a good product',
              price: 230,
            },
            {
              productname: 'shoe lace2',
              quantity: 33,
              minimum: 4,
              description: 'a good product',
              price: 230,
            },
          ],
        },
      };

      const component = shallow(<ProductComponent {...props} />);
    });

    it('should render the product container', () => {
      const component = shallow(<ProductContainer />);
      expect(component.exists()).toBe(true);
      expect(component).toMatchSnapshot();
    });
  });

  it('should ensures mapDispatchToProps dispatches the specified actions', () => {
    const mock = configureStore([thunk]);
    const store = mock(props);
    const component = shallow(
      <Provider store={store}>
        <ProductContainer />
      </Provider>,
    );
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).fetchAllProducts();
    mapDispatchToProps(dispatch).searchProduct();
    mapDispatchToProps(dispatch).addToCart();
    mapDispatchToProps(dispatch).removeFromCart();
  });
});

describe('TEST FOR GET PRODUCT', () => {
  const payload = {
    data: {
      productname: 'productname',
      id: 1,
      quantity: 1,
      minimum: 2,
    },
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore({});
  let wrapper;
  const response = { message: 'fetchall product' };
  axios.get.mockResolvedValue(response);
  store.dispatch(fetchAllProducts()).then(() => {
    store.dispatch(setAllProducts(payload));
  });
});

describe('TEST FOR SEARCH PRODUCT', () => {
  const mockStore = configureStore([thunk]);
  const store = mockStore({});
  let wrapper;
  const response = { message: 'delete product' };
  axios.get.mockResolvedValue(response);
  store.dispatch(searchProduct()).then(() => {});
});

describe('Products Reducers', () => {
  it('should setup default state values', () => {
    const state = productReducer(undefined, {
      type: '@@INIT',
    });
    expect(state).toEqual({});
  });

  it('it should change product payload', () => {
    const action = {
      type: types.LOAD_ALL_PRODUCT,
      payload: 'products',
    };
    const state = productReducer(undefined, action);
    expect(state.payload).toEqual(action.payload);
  });

  it('it should change get a product payload', () => {
    const action = {
      type: types.GET_A_PRODUCT,
      product: 'snickers',
    };
    const state = productReducer(undefined, action);
    expect(state.product).toEqual(action.product);
  });
});

describe('Product Actions', () => {
  it('it should set the all product state', () => {
    const action = setAllProducts('products');
    expect(action).toEqual({
      type: types.LOAD_ALL_PRODUCT,
      payload: 'products',
    });
  });

  it('it should set the get a Product state', () => {
    const action = getAProducts('timberland');
    expect(action).toEqual({
      type: types.GET_A_PRODUCT,
      product: 'timberland',
    });
  });

  it('it should set the set remove from cart state', () => {
    const action = removeFromCart('timberland');
    expect(action).toEqual({
      type: types.REMOVE_FROM_CART,
      item: 'timberland',
    });
  });

  it('it should set item to be added to cart', () => {
    const action = addToCart('timberland');
    expect(action).toEqual({
      type: types.ADD_TO_CART,
      item: 'timberland',
    });
  });
});
