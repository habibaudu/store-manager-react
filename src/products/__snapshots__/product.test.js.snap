// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PRODUCT TEST SUITE Product Component should render the PRODUCT Page 1`] = `
<div>
  <SlideShowComponent />
  <div
    className="mobile_header"
  >
    <img
      src={Object {}}
    />
  </div>
  <section
    id="title"
  >
    Search for Product
    <form
      className="byid"
      id="search"
      onSubmit={[Function]}
    >
      <input
        id="filter2"
        name="search"
        placeholder="Enter the exact name product  eg Men Loafers"
        type="text"
      />
      <button
        type="submit"
      >
        search 
      </button>
    </form>
  </section>
   
  <section
    id="title"
  >
    <Link
      replace={false}
      to="/cart"
    >
      view cart
    </Link>
  </section>
  <div
    className="article clearfix"
    id="allproduct"
  >
    <h3>
      loading product please wait...
    </h3>
  </div>
</div>
`;

exports[`PRODUCT TEST SUITE Product Component should render the product container 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
